src/lua/datetime.lua

```diff
@@ -864,6 +880,85 @@ local function datetime_interval_add(lhs, rhs)
     end
 end
 
+--[[
+    Parse partial ISO-8601 date string
+
+    Accepetd formats are:
+
+    Basic      Extended
+    20121224   2012-12-24   Calendar date   (ISO 8601)
+    2012359    2012-359     Ordinal date    (ISO 8601)
+    2012W521   2012-W52-1   Week date       (ISO 8601)
+    2012Q485   2012-Q4-85   Quarter date
+
+    Returns pair of constructed datetime object, and length of string
+    which has been accepted by parser.
+]]
+local function datetime_parse_date(str)
+    check_str("datetime.parse_date()")
+    local dt = ffi.new('dt_t[1]')
+    local len = tonumber(builtin.tnt_dt_parse_iso_date(str, #str, dt))
+    if len == 0 then
+        error(('invalid date format %s'):format(str), 2)
+    end
+    return datetime_new_dt(dt[0]), len
```

Comment:
To be honest it looks a bit inconsistent.

```
tarantool> dt.parse_date('2021-01-01')
---
- 2021-01-01T00:00:00Z
- 10
...

tarantool> dt.parse('2021-01-01')
---
- 2021-01-01T00:00:00Z
...
```

I believe both parsing functions should return equal number of arguments.

Reply:
`date.parse_date` was actually extension to RFC and since in the test we needed to know which part of input string has been accepted (i.e. allowing only prefix to be accepted) thus I'm returning second value - length.

While `date.parse` now returns single value, but internally allowing same algorithm of partial string parsing. Now you've asked it I believe that they both should return additional length, to easier scenario of parsing datetime in the longer string with additional content.

Will add.

Reply:
```diff
diff --git a/src/lib/core/datetime.c b/src/lib/core/datetime.c
index 225045cf8..081edf711 100644
--- a/src/lib/core/datetime.c
+++ b/src/lib/core/datetime.c
@@ -101,14 +101,15 @@ tm_to_datetime(struct tnt_tm *tm, struct datetime *date)
 	date->tzoffset = tm->tm_gmtoff / 60;
 }
 
-char *
+size_t
 datetime_strptime(struct datetime *date, const char *buf, const char *fmt)
 {
 	struct tnt_tm t = { .tm_epoch = 0 };
 	char * ret = tnt_strptime(buf, fmt, &t);
-	if (ret != NULL)
-		tm_to_datetime(&t, date);
-	return ret;
+	if (ret == NULL)
+		return 0;
+	tm_to_datetime(&t, date);
+	return ret - buf;
 }
 
 void
diff --git a/src/lib/core/datetime.h b/src/lib/core/datetime.h
index 2420fd135..200edc931 100644
--- a/src/lib/core/datetime.h
+++ b/src/lib/core/datetime.h
@@ -136,11 +136,11 @@ datetime_parse_full(struct datetime *date, const char *str, size_t len,
  * @param date output datetime value
  * @param buf input text buffer (0-terminated)
  * @param fmt format to use for parsing
- * @retval Upon successful completion returns a pointer to the character
- *         following the last character parsed. Otherwise NULL
+ * @retval * @retval Upon successful completion returns length of accepted
+ *         prefix substring. Otherwise 0
  * @sa strptime()
  */
-char *
+size_t
 datetime_strptime(struct datetime *date, const char *buf, const char *fmt);
 
 #if defined(__cplusplus)
diff --git a/src/lua/datetime.lua b/src/lua/datetime.lua
index 8bf29054a..0a01418c4 100644
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -74,9 +74,9 @@ size_t tnt_datetime_to_string(const struct datetime * date, char *buf,
                               ssize_t len);
 size_t tnt_datetime_strftime(const struct datetime *date, char *buf,
                              uint32_t len, const char *fmt);
-bool   tnt_datetime_parse_full(struct datetime *date, const char *str,
+size_t tnt_datetime_parse_full(struct datetime *date, const char *str,
                                size_t len, int32_t offset);
-char * tnt_datetime_strptime(struct datetime *date, const char *buf,
+size_t tnt_datetime_strptime(struct datetime *date, const char *buf,
                              const char *fmt);
 void   tnt_datetime_now(struct datetime *now);
 
@@ -907,28 +907,29 @@ end
     datetime parse function for strings in extended iso-8601 format
     assumes to deal with date T time time_zone at once
        date [T] time [ ] time_zone
-    Returns constructed datetime object.
+    Returns constructed datetime object and length of accepted string.
 ]]
 local function datetime_parse_full(str, tzoffset)
     check_str(str, "datetime.parse()")
     local date = ffi.new(datetime_t)
-    local rc = builtin.tnt_datetime_parse_full(date, str, #str, tzoffset)
-    if rc == false then
+    local len = builtin.tnt_datetime_parse_full(date, str, #str, tzoffset)
+    if len == 0 then
         error(("Could not parse '%s'"):format(str))
     end
-    return date
+    return date, len
 end
 
 --[[
-    Parse datetime string given `strptime` like format
+    Parse datetime string given `strptime` like format.
+    Returns constructed datetime object and length of accepted string.
 ]]
 local function datetime_parse_format(str, fmt)
     local date = ffi.new(datetime_t)
-    local res = builtin.tnt_datetime_strptime(date, str, fmt)
-    if res == box.NULL then
+    local len = builtin.tnt_datetime_strptime(date, str, fmt)
+    if len == 0 then
         error(("Could not parse '%s' using '%s' format"):format(str, fmt))
     end
-    return date
+    return date, len
 end
 
 local function datetime_parse_from(str, obj)
diff --git a/src/lua/tnt_datetime.c b/src/lua/tnt_datetime.c
index 51869a5a1..49fa1b130 100644
--- a/src/lua/tnt_datetime.c
+++ b/src/lua/tnt_datetime.c
@@ -14,7 +14,7 @@ tnt_datetime_strftime(const struct datetime *date, char *buf, size_t len,
 	return datetime_strftime(date, buf, len, fmt);
 }
 
-char *
+size_t
 tnt_datetime_strptime(struct datetime *date, const char *buf, const char *fmt)
 {
 	return datetime_strptime(date, buf, fmt);
```